pkg/user/signer.go

```diff
@@ -0,0 +1,360 @@
+package user
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/celestiaorg/celestia-app/app/encoding"
+	blob "github.com/celestiaorg/celestia-app/x/blob/types"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/grpc/tmservice"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"google.golang.org/grpc"
+)
+
+const defaultPollTime = 3 * time.Second
+
+// Signer is an abstraction for managing building, signing and broadcasting Celestia transactions
+type Signer struct {
+	keys          keyring.Keyring
+	address       sdktypes.Address
+	encCfg        encoding.Config
+	grpc          *grpc.ClientConn
+	pk            cryptotypes.PubKey
+	chainID       string
+	accountNumber uint64
+	pollTime      time.Duration
+
+	mtx                   sync.RWMutex
+	lastSignedSequence    uint64
+	lastConfirmedSequence uint64
+}
+
+// NewSigner returns a new signer using the provided keyring
+func NewSigner(
+	keys keyring.Keyring,
+	conn *grpc.ClientConn,
+	address sdktypes.Address,
+	encCfg encoding.Config,
+	chainID string,
+	accountNumber uint64,
+	sequence uint64,
+) (*Signer, error) {
+	// check that the address exists
+	record, err := keys.KeyByAddress(address)
+	if err != nil {
+		return nil, err
+	}
+
+	pk, err := record.GetPubKey()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Signer{
+		keys:                  keys,
+		address:               address,
+		grpc:                  conn,
+		encCfg:                encCfg,
+		pk:                    pk,
+		chainID:               chainID,
+		accountNumber:         accountNumber,
+		lastSignedSequence:    sequence,
+		lastConfirmedSequence: sequence,
+		pollTime:              defaultPollTime,
+	}, nil
+}
+
+// SetupSingleSigner relies on the keyring with a single key. It extracts the address from the key and uses the
+// grpc connection to populate the chainID, account number and sequence number.
```

Comment:
[optional] I'm confused by this usage of "relies". Perhaps clearer:

```suggestion
// SetupSingleSigner sets up a signer based on the provided keyring. The keyring
// must contain exactly one key. It extracts the address from the key and uses
// the grpc connection to populate the chainID, account number, and sequence
// number.
```